Source/DafnyRuntime/DafnyRuntimeRust/src/lib.rs

```diff
@@ -9,6 +16,346 @@ pub use num::NumCast;
+                array.push(value.clone());
+              },
+            Sequence::ConcatSequence { left, right, .. } =>
+              // Let's create an array of size length and fill it up recursively
+              {
+                Sequence::<T>::append_recursive(array, left);
+                Sequence::<T>::append_recursive(array, right);
+              }
+            Sequence::LazySequence { boxed, .. } =>
+            Sequence::<T>::append_recursive(array, boxed.borrow().as_ref()),
+        }
+    }
+
+    fn node_count(&self) -> usize {
+        match self {
+            Sequence::ArraySequence { node_count, .. } =>
+              // The length of the elements
+              *node_count,
+            Sequence::ConcatSequence { node_count, ..  } =>
+              *node_count,
+            Sequence::LazySequence { node_count, .. } =>
+              *node_count,
+        }
+    
+    }
+    /// Returns the cardinality of this [`Sequence<T>`].
+    // The cardinality returns the length of the sequence
+    fn cardinality(&self) -> SizeT {
+        match self {
+            Sequence::ArraySequence { values, .. } =>
+              // The length of the elements
+              values.len(),
+            Sequence::ConcatSequence { length, .. } =>
+              *length,
+            Sequence::LazySequence { length, .. } =>
+              *length,
+        }
+    }
+    fn select(&self, index: SizeT) -> T {
+        let array = self.to_array();
+        array[index].clone()
+    }
+    
+}
+
+// **************
+// Immutable maps
+// **************
+
+#[allow(dead_code)]
+struct Map<K, V> 
+  where K: Clone + Eq + std::hash::Hash, V: Clone
+{
+    data: Rc<Sequence<(K, V)>>,
+    // Any time we explicitly access this map, we index the data
+    cache: RefCell<Option<HashMap<K, V>>>
+}
+
+#[allow(dead_code)]
+impl <K, V> Map<K, V>
+  where K: Clone + Eq + std::hash::Hash, V: Clone
+{
+    fn new_empty() -> Rc<Map<K, V>> {
+        Rc::new(Map {
+            data: Sequence::new_array_sequence(&Rc::new(Vec::new())),
+            cache: RefCell::new(None)
+        })
+    }
+    fn new_from_sequence(data: &Rc<Sequence<(K, V)>>) -> Rc<Map<K, V>> {
+        Rc::new(Map {
+            data: Rc::clone(data),
+            cache: RefCell::new(None)
+        })
+    }
+    fn to_hashmap_owned<K2, V2>(&self, converter_k: fn(&K)->K2, converter_v: fn(&V)->V2) -> HashMap<K2, V2>
+      where K2: Eq + std::hash::Hash, V2: Clone
+    {
+        let mut result: HashMap<K2, V2> = HashMap::new();
+        for (k, v) in self.data.to_array().iter() {
+            result.insert(converter_k(k), converter_v(v));
+        }
+        result
+    }
+    fn compute_hashmap(&self) {
+        let mut cache = self.cache.borrow_mut();
+        if cache.is_none() {
+            *cache = Some(self.to_hashmap_owned(
+                |x| x.clone(),
+                |x| x.clone()
+            ));
+        }
+    }
+    // Dafny will normally guarantee that the key exists.
```

Comment:
Which means the `unwrap()`s should never panic, right?

Reply:
Correct!